tests/catch/unit/memory/hipFree.cc

```diff
@@ -0,0 +1,452 @@
+    default:
+      return hipErrorIllegalState;
+      break;
+  }
+}
+
+template <> hipError_t freeStuff<hiparray>(CUarray ptr, FreeType type) {
+  switch (type) {
+    case ArrayFree:
+      return hipErrorIllegalState;
+      break;
+    case ArrayDestroy:
+      return hipArrayDestroy(ptr);
+      break;
+    default:
+      return hipErrorIllegalState;
+      break;
+  }
+}
+#else
+template <> hipError_t freeStuff<hipArray_t>(hipArray_t ptr, FreeType type) {
+  switch (type) {
+    case ArrayFree:
+      return hipFreeArray(ptr);
+      break;
+    case ArrayDestroy:
+      return hipArrayDestroy(ptr);
+      break;
+    default:
+      return hipErrorIllegalState;
+      break;
+  }
+}
+#endif
+
+// Helper function to check if work on device is done
+template <typename T> hipError_t workIsDoneCheck(T ptr, FreeType fType) {
+  auto error = hipSuccess;
+  // free memory
+  error = freeStuff(ptr, fType);
+  if (error != hipSuccess) {
+    printf("Breaking on first free \n");
+    return error;
+  }
+  // ensure memory is freed
+  // SegFaults the when using hipFreeArray or hipArrayDestroy  </3
+  error = freeStuff(ptr, fType);
+  if (error != hipErrorInvalidValue) {
+    printf("Breaking on second free \n");
+    return error;
+  }
+  // verify synchronization
+  error = hipStreamQuery(nullptr);
+  if (error != hipSuccess) {
+    printf("Breaking on hipStreamQuery free \n");
+  }
+
+  return error;
+}
+
+// DevFree, ArrayFree, ArrayDestroy, HostFree
+TEMPLATE_TEST_CASE("Unit_hipFreeImplicitSyncDev", "", char, float, float2, float4) {
+  enum StreamType { NULLSTR, CREATEDSTR };
+  auto streamType = GENERATE(NULLSTR, CREATEDSTR);
+  hipStream_t stream{nullptr};
+  if (streamType == CREATEDSTR) HIP_CHECK(hipStreamCreate(&stream));
+
+  TestType* devPtr{};
+  size_t size_mult = GENERATE(1, 32, 64, 128, 256);
+  HIP_CHECK(hipMalloc(&devPtr, sizeof(TestType) * size_mult));
+
+  lauchLongRunningKernel(50);
+  // make sure device is busy
+  HIP_CHECK_ERROR(hipStreamQuery(nullptr), hipErrorNotReady);
+  HIP_CHECK(workIsDoneCheck<TestType*>(devPtr, DevFree));
+  if (streamType == CREATEDSTR) HIP_CHECK(hipStreamDestroy(stream));
+}
+
+TEMPLATE_TEST_CASE("Unit_hipFreeImplicitSyncHost", "", char, float, float2, float4) {
+  enum StreamType { NULLSTR, CREATEDSTR };
+  auto streamType = GENERATE(NULLSTR, CREATEDSTR);
+  hipStream_t stream{nullptr};
+  if (streamType == CREATEDSTR) HIP_CHECK(hipStreamCreate(&stream));
+  TestType* hostPtr{};
+  size_t size_mult = GENERATE(1, 32, 64, 128, 256);
+
+  HIP_CHECK(hipHostMalloc(&hostPtr, sizeof(TestType) * size_mult));
+
+  lauchLongRunningKernel(50);
+  // make sure device is busy
+  HIP_CHECK_ERROR(hipStreamQuery(nullptr), hipErrorNotReady);
+  HIP_CHECK(workIsDoneCheck<TestType*>(hostPtr, HostFree));
+
+  if (streamType == CREATEDSTR) HIP_CHECK(hipStreamDestroy(stream));
+}
+
+TEMPLATE_TEST_CASE("Unit_hipFreeImplicitSyncArray", "", char, float, float2, float4) {
+  HipTest::HIP_SKIP_TEST("EXSWCPHIPT-79");
```

Comment:
Doers this work with any platform?

Reply:
On AMD devices the test causes a segfault, while on Nvidia devices the behaviour is handled however we still get error code 709: (context is destroyed)

Reply:
what returns that error on NV? Seems like they're just not crashing but still returning some garbage since this has nothing to do with context.

Reply:
The errors have been removed by patch in the HIPAMD repo: 06f64e1a53ddc9b0ca02993ff5bff95bea1f8f7f